File: Medium/63. Unique Paths II/solution.java
/**
 * TC: O(n + m)
 * SC: O(mn)
 */
class Solution {
    public int uniquePathsWithObstacles(int[][] obstacleGrid) {
        
        if (obstacleGrid[0][0] == 1) {
            return 0;
        }
        
        int rows = obstacleGrid.length;
        int cols = obstacleGrid[0].length;
        int[][] moves = new int[rows][cols];
        
        int i = 0;
        for (;i < cols; i++) {
            if (obstacleGrid[0][i] == 1) {
                moves[0][i] = 0;
                break;
            } else {
                moves[0][i] = 1;
            }
        }
        while (i < cols) {
            moves[0][i++] = 0;
        }
        i = 0;
        for (; i < rows; i++) {
            if (obstacleGrid[i][0] == 1) {
                moves[i][0] = 0;
                break;
            } else {
                moves[i][0] = 1;
            }
            
        }
        while (i < rows) {
            moves[i++][0] = 0;
        }
        for (i = 1; i < rows; i++) {
            for (int j = 1; j < cols; j++) {
                if (obstacleGrid[i][j] == 1) {
                    moves[i][j] = 0;
                } else {
                    moves[i][j] = moves[i - 1][j] + moves[i][j - 1];
                }
            }
        }
        return moves[rows - 1][cols - 1];
    }
}

File: Medium/63. Unique Paths II/solution.py

# TC: O(n*m); SC: O(m * n)
class Solution:
    def uniquePathsWithObstacles(self, obstacleGrid: List[List[int]]) -> int:
        
        if not obstacleGrid or len(obstacleGrid) == 0 or len(obstacleGrid[0]) == 0:
            return 0
        
        if obstacleGrid[0][0] == 1:
            return 0
        rows = len(obstacleGrid)
        cols = len(obstacleGrid[0])
        
        dp = [[0 for i in range(cols)] for j in range(rows)]
        
        i = 0
        isobstacle = False
        for i in range(cols):
            if obstacleGrid[0][i] == 1:
                isobstacle = True
                
            if isobstacle:
                dp[0][i] = 0
            else:
                dp[0][i] = 1
                
        
        print (dp)
        i = 0
        isobstacle = False
        for i in range(rows):
            if obstacleGrid[i][0] == 1:
                isobstacle = True
                
            if isobstacle:
                dp[i][0] = 0
            else:
                dp[i][0] = 1
        
        
        for i in range(1,rows):
            for j in range(1, cols):
                if obstacleGrid[i][j] == 1:
                    dp[i][j] = 0
                else:
                    dp[i][j] = dp[i - 1][j] + dp[i][j - 1]
        
        return dp[rows - 1][cols - 1]
        
        